File: app/src/main/java/com/android/oleksandrpriadko/demo/MainActivity.java
package com.android.oleksandrpriadko.demo;

import com.android.oleksandrpriadko.demo.loggalitic.LoggaliticDemoActivity;
import com.crashlytics.android.answers.Answers;
import io.fabric.sdk.android.Fabric;
import java.util.ArrayList;
import java.util.List;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.android.oleksandrpriadko.demo.adapter.AdapterDemos;
import com.android.oleksandrpriadko.demo.adapter.Demo;
import com.android.oleksandrpriadko.demo.item_decoration.ItemDecorationActivity;
import com.android.oleksandrpriadko.recycler_adapter.BaseItemListener;
import com.android.oleksandrpriadko.recycler_decoration.ItemDecorationMargin;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity extends AppCompatActivity {

    @BindView(R.id.recView)
    RecyclerView mRecyclerViewDemos;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fabric.with(this, new Answers());
        setContentView(R.layout.activity_main);

        ButterKnife.bind(this);

        intRecViewDemos();
    }

    private void intRecViewDemos() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(this);
        layoutManager.setOrientation(LinearLayoutManager.VERTICAL);
        mRecyclerViewDemos.setLayoutManager(layoutManager);

        AdapterDemos adapterDemos = createAdapter();
        mRecyclerViewDemos.setAdapter(adapterDemos);

        ItemDecorationMargin itemDecorationMargin = new ItemDecorationMargin(
            30,
            true,
            true,
            layoutManager.getOrientation());
        mRecyclerViewDemos.addItemDecoration(itemDecorationMargin);
    }

    private AdapterDemos createAdapter() {
        BaseItemListener<Demo> baseItemListener = new BaseItemListener<Demo>() {
            @Override
            public void isEmpty(boolean isEmpty) {

            }

            @Override
            public void itemClicked(int position, Demo item) {
                Intent intent = new Intent(MainActivity.this, item.getClazz());
                startActivity(intent);
            }
        };
        AdapterDemos adapterDemos = new AdapterDemos(baseItemListener);
        List<Demo> demoList = new ArrayList<>();
        Demo demo = new Demo(
            ItemDecorationActivity.class,
            "ItemDecoration",
            null);
        demoList.add(demo);

        demo = new Demo(
            LoggaliticDemoActivity.class,
            "Loggalitic",
            null);
        demoList.add(demo);

        adapterDemos.setData(demoList);

        return adapterDemos;
    }
}

File: app/src/main/java/com/android/oleksandrpriadko/demo/main/MainActivity.kt
package com.android.oleksandrpriadko.demo.main

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.android.oleksandrpriadko.demo.R
import com.android.oleksandrpriadko.demo.main.adapter.AdapterDemos
import com.android.oleksandrpriadko.demo.main.adapter.Demo
import com.android.oleksandrpriadko.demo.item_decoration.ItemDecorationActivity
import com.android.oleksandrpriadko.demo.logpublish.LogPublishActivity
import com.android.oleksandrpriadko.extension.dimenPixelSize
import com.android.oleksandrpriadko.recycler_adapter.BaseItemListener
import com.android.oleksandrpriadko.recycler_decoration.ItemDecorationVerticalGridMargin
import com.crashlytics.android.answers.Answers
import io.fabric.sdk.android.Fabric
import kotlinx.android.synthetic.main.main_activity_main.*
import java.util.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Fabric.with(this, Answers())
        setContentView(R.layout.main_activity_main)

        intRecViewDemos()
    }

    private fun intRecViewDemos() {
        val spanCount = 2
        val layoutManager = GridLayoutManager(this, spanCount)
        layoutManager.orientation = RecyclerView.VERTICAL
        recyclerView.layoutManager = layoutManager

        val adapterDemos = createAdapter()
        recyclerView.adapter = adapterDemos

        val itemDecorationMargin = ItemDecorationVerticalGridMargin(
                spanCount = spanCount,
                left = dimenPixelSize(R.dimen.main_margin_left_right_card),
                top = dimenPixelSize(R.dimen.main_margin_top_bottom_card),
                right = dimenPixelSize(R.dimen.main_margin_left_right_card),
                bottom = dimenPixelSize(R.dimen.main_margin_top_bottom_card),
                includeEdge = true
        )
        recyclerView.addItemDecoration(itemDecorationMargin)
    }

    private fun createAdapter(): AdapterDemos {
        val baseItemListener = object : BaseItemListener<Demo> {
            override fun isEmpty(isEmpty: Boolean) {

            }

            override fun itemClicked(position: Int, item: Demo) {
                val intent = Intent(this@MainActivity, item.clazz)
                startActivity(intent)
            }
        }
        val adapterDemos = AdapterDemos(baseItemListener)
        val demoList = ArrayList<Demo>()
        var demo = Demo(
                ItemDecorationActivity::class.java,
                "ItemDecoration",
                null,
                R.drawable.main_ic_workflow_512)
        demoList.add(demo)

        demo = Demo(
                LogPublishActivity::class.java,
                "Analytics",
                null,
                R.drawable.main_ic_analytics_512)
        demoList.add(demo)

        adapterDemos.setData(demoList)

        return adapterDemos
    }
}